components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java

```diff
@@ -111,12 +114,24 @@ private void handleNotificationMessage(String eventType, long timestamp, String
             gatewayEvent.getGatewayLabels().retainAll(gatewayArtifactSynchronizerProperties.getGatewayLabels());
             if (!gatewayEvent.getGatewayLabels().isEmpty()) {
                 String gatewayLabel = gatewayEvent.getGatewayLabels().iterator().next();
+                Runnable task = null;
                 if (APIConstants.EventType.DEPLOY_API_IN_GATEWAY.name().equals(eventType)) {
-                    inMemoryApiDeployer.deployAPI(gatewayEvent.getApiId(), gatewayLabel);
+                    task = new Runnable() {
+                        @Override
+                        public void run() {
+                            inMemoryApiDeployer.deployAPI(gatewayEvent.getApiId(), gatewayLabel);
+                        }
+                    };
                 } else if (APIConstants.EventType.REMOVE_API_FROM_GATEWAY.name().equals(eventType)) {
-                    inMemoryApiDeployer.unDeployAPI(gatewayEvent.getApiId(), gatewayLabel);
-                }
+                    task = new Runnable() {
+                        @Override
+                        public void run() {
+                            inMemoryApiDeployer.unDeployAPI(gatewayEvent.getApiId(), gatewayLabel);
+                        }
+                    };
                 }
+                scheduler.schedule(task, 1, TimeUnit.MILLISECONDS);
```

Comment:
Add a meaningful debug log correlating EventId, APIName and other useful details in the event.

Reply:
https://github.com/wso2/carbon-apimgt/pull/9016/commits/bb95666b85e9ca2fe650f54b212bcdce5d06008f